zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/X509ZNodeGroupAclProvider.java

```diff
@@ -0,0 +1,156 @@
+import javax.net.ssl.X509TrustManager;
+import javax.security.auth.x500.X500Principal;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.common.ZKConfig;
+import org.apache.zookeeper.data.Id;
+import org.apache.zookeeper.server.ServerCnxn;
+import org.apache.zookeeper.server.auth.ServerAuthenticationProvider;
+import org.apache.zookeeper.server.auth.X509AuthenticationUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A ServerAuthenticationProvider implementation that does both authentication and authorization for protecting znodes from unauthorized access.
+ * Znodes are grouped into domains according to their ownership, and clients are granted access permission to domains.
+ * Authentication mechanism is same as in X509AuthenticationProvider.
+ * Authorization is done by checking with clients' URI (uniform resource identifier) in ACL metadata for matched domains.
+ * This class is meant to support the following use patterns:
+ *  1.Multiple creators, one reader - ZNodes, created by multiple identities from multiple domains,
+ *    need to be read by an identity from a different domain.
+ *  2.Multiple creators, multiple readers - ZNodes, created by multiple identities from multiple domains,
+ *    need to be read by multiple identities from multiple domains.
+ *  3.Multiple creators, one all-permission accessor - ZNodes, created by multiple identities from multiple domains,
+ *    need to be read/update/deleted by an identity from a different domain.
+ * Optional features include:
+ *    set client Id as znode ACL when creating nodes,
+ *    add additional public read ACL to znodes based on path.
+ */
+public class X509ZNodeGroupAclProvider extends ServerAuthenticationProvider {
+
+  private static final Logger LOG = LoggerFactory.getLogger(X509ZNodeGroupAclProvider.class);
+  private final String logStrPrefix = this.getClass().getName() + ":: ";
+  private final X509TrustManager trustManager;
+  private final X509KeyManager keyManager;
+  static final String ZOOKEEPER_ZNODEGROUPACL_SUPERUSER = "zookeeper.znodeGroupAcl.superUser";
+
+  public X509ZNodeGroupAclProvider() {
+    ZKConfig config = new ZKConfig();
+    this.keyManager = X509AuthenticationUtil.createKeyManager(config);
+    this.trustManager = X509AuthenticationUtil.createTrustManager(config);
+  }
+
+  public X509ZNodeGroupAclProvider(X509TrustManager trustManager, X509KeyManager keyManager) {
+    this.trustManager = trustManager;
+    this.keyManager = keyManager;
+  }
+
+  @Override
+  public KeeperException.Code handleAuthentication(ServerObjs serverObjs, byte[] authData) {
+    ServerCnxn cnxn = serverObjs.getCnxn();
+    X509Certificate clientCert;
+    try {
+      clientCert = X509AuthenticationUtil.getAndAuthenticateClientCert(cnxn, trustManager);
+    } catch (KeeperException.AuthFailedException e) {
+      return KeeperException.Code.AUTHFAILED;
+    }
+
+    // Extract URI from certificate
```

Comment:
Strictly speaking, this is not a URI - this is just extracting whatever the server wants to extract based on the regex set in the config, correct? So we should capture that here.

Reply:
In this case let's call it clientId, and for our use case, we use URI as clientId.

Reply:
I'm not saying we should just change it back to `clientId`. I'm saying that we should give ample context. Either way is acceptable, but we can put some comments here to help the reader understand that whatever it is called, the expected usage pattern is that this string will be some kind of an URI that can be mapped to a domain. And that user has the flexibility to extract an appropriate string from the certificate using SAN field and regex matches.